packages/shared/src/permissions/permissions.utils.ts

```diff
@@ -31,3 +31,77 @@ export function hasPermission(
 export function roleSupportsEnvLimit(role: MemberRole): boolean {
   return ["engineer", "experimenter"].includes(role);
 }
+
+export type ReadAccessFilter = {
+  globalReadAccess: boolean;
+  projects: { id: string; readAccess: boolean }[];
+};
+
+export function getReadAccessFilter(userPermissions: UserPermissions) {
+  const readAccess: ReadAccessFilter = {
+    globalReadAccess: userPermissions.global.permissions.readData || false,
+    projects: [],
+  };
+
+  Object.keys(userPermissions.projects).forEach((project) => {
+    readAccess.projects.push({
+      id: project,
+      readAccess:
+        userPermissions.projects[project].permissions.readData || false,
+    });
+  });
+
+  return readAccess;
+}
+
+export function hasReadAccess(
+  readAccessFilter: ReadAccessFilter,
+  resourceProjects: string[] // feature.project or experiment.project or metric.projects
+): boolean {
+  if (readAccessFilter.projects.length === 0 || resourceProjects.length === 0) {
+    return readAccessFilter.globalReadAccess;
+  }
+
+  if (readAccessFilter.globalReadAccess) {
+    let userHasProjectSpecificAccessForEachResourceProject = true;
+
+    // Check if user has project specific access for each project the resources is in
+    for (let i = 0; i < resourceProjects.length; i++) {
+      const projectAccessIndex = readAccessFilter.projects.findIndex(
+        (projectAccess) => projectAccess.id === resourceProjects[i]
+      );
+      if (projectAccessIndex === -1) {
+        userHasProjectSpecificAccessForEachResourceProject = false;
+        break;
+      }
+    }
+    // if user has project specific role for each of the resource's projects, only allow readaccess if they have read access for every project the resource is in.
+    // e.g. metric.projects = ["project1", "project2", "project3"] and readAccessFilter.projects = [{id: "project1", readAccess: true}, {id: "project2", readAccess: true}, {id: "project3", readAccess: true}]
```

Comment:
Is this comment correct?

If a metric is tied to project A and B, I would think you only need access to at least one of the projects to read the metric.  This comment suggests you would need read access to both.